File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque

way_cnt = 0
def solution(numbers, target):
    global way_cnt
    N = len(numbers)

    def dfs(operator, depth):
        global way_cnt
        
        if depth == N-1:
            tmp_sum = 0
            for i in range(N):
                if operator[i] == '+':
                    tmp_sum += numbers[i]
                else:
                    tmp_sum -= numbers[i]
            if tmp_sum == target:
                way_cnt += 1
            return
        
        dfs(operator + '+', depth+1)
        dfs(operator + '-', depth+1)
             
    dfs('+', 0)
    dfs('-', 0)
    return way_cnt

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.cpp
#include <iostream>
#include <vector>
using namespace std;

int way_cnt = 0;

void dfs(string operatorStr, int depth, const vector<int>& numbers, int target) {
    if (depth == numbers.size() - 1) {
        int tmp_sum = 0;
        for (int i = 0; i < numbers.size(); ++i) {
            if (operatorStr[i] == '+') {
                tmp_sum += numbers[i];
            } else {
                tmp_sum -= numbers[i];
            }
        }
        if (tmp_sum == target) {
            way_cnt++;
        }
        return;
    }

    dfs(operatorStr + '+', depth + 1, numbers, target);
    dfs(operatorStr + '-', depth + 1, numbers, target);
}

int solution(vector<int> numbers, int target) {
    dfs("+", 0, numbers, target);
    dfs("-", 0, numbers, target);
    return way_cnt;
}